File: January2022/Java/SumofRootToLeafBinaryNumbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class SumofRootToLeafBinaryNumbers {
    int totalSum = 0;
    public int sumRootToLeaf(TreeNode root) {
        if(root==null){
            return 0;
        }

        postOrderTraversal(root, 0);
        return totalSum;
    }
    // O(n), where n is the no of nodes in the tree
    private void postOrderTraversal(TreeNode root, int currVal){
        if(root==null){
            return ;
        }
        currVal = currVal + root.val;

        // leaf node or not
        if(root.left == null && root.right==null){
            totalSum += currVal;
            return ;
        }
        postOrderTraversal(root.left , currVal* 2);
        postOrderTraversal(root.right , currVal* 2);

    }
}

File: January2022/Python/SumOfRootToLeafBinaryNumbers.py
# Author :  Karthik K (https://www.linkedin.com/in/karthikmunirathinam/)

# TC : O(n) where,n-->Number of nodes
# SC : O(n) Recursive stack space

class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        def dfs(node, currentVal):
            nonlocal totalSum
            if not node:
                return
            currentVal = currentVal*2+node.val
            if not node.left and not node.right:
                totalSum += currentVal
            dfs(node.left, currentVal)
            dfs(node.right, currentVal)
        totalSum = 0
        dfs(root, 0)
        return totalSum
